ogr/ogrsf_frmts/miramon/ogrmiramonlayer.cpp

```diff
@@ -0,0 +1,2613 @@
+            hMMFeature.pRecords[0].pField[iField].bIsValid = 1;
+        }
+        else
+        {
+            CPLError(CE_Failure, CPLE_NotSupported,
+                     "MiraMon: Field type %d not processed by MiraMon\n",
+                     eFType);
+            return OGRERR_UNSUPPORTED_GEOMETRY_TYPE;
+        }
+    }
+
+    return OGRERR_NONE;
+}
+
+/****************************************************************************/
+/*                             GetLayerDefn()                               */
+/*                                                                          */
+/****************************************************************************/
+OGRFeatureDefn *OGRMiraMonLayer::GetLayerDefn()
+{
+    return poFeatureDefn;
+}
+
+/****************************************************************************/
+/*                             GetExtent()                                  */
+/*                                                                          */
+/*      Fetch extent of the data currently stored in the dataset.           */
+/*      The bForce flag has no effect on SHO files since that value         */
+/*      is always in the header.                                            */
+/****************************************************************************/
+
+OGRErr OGRMiraMonLayer::GetExtent(OGREnvelope *psExtent, int bForce)
+
+{
+    if (phMiraMonLayer)
+    {
+        psExtent->MinX = phMiraMonLayer->TopHeader.hBB.dfMinX;
+        psExtent->MaxX = phMiraMonLayer->TopHeader.hBB.dfMaxX;
+        psExtent->MinY = phMiraMonLayer->TopHeader.hBB.dfMinY;
+        psExtent->MaxY = phMiraMonLayer->TopHeader.hBB.dfMaxY;
+    }
+    else
+    {
+        if (!bForce)
+            return OGRERR_FAILURE;
+    }
+
+    return OGRERR_NONE;
+}
+
+/****************************************************************************/
+/*                           TestCapability()                               */
+/****************************************************************************/
+
+int OGRMiraMonLayer::TestCapability(const char *pszCap)
+
+{
+    if (EQUAL(pszCap, OLCRandomRead))
+        return TRUE;
+
+    if (EQUAL(pszCap, OLCSequentialWrite))
+        return TRUE;
+
+    if (EQUAL(pszCap, OLCFastFeatureCount))
+        return TRUE;
+
+    if (EQUAL(pszCap, OLCFastGetExtent))
+        return TRUE;
+
+    if (EQUAL(pszCap, OLCCreateField))
+        return TRUE;
+
+    if (EQUAL(pszCap, OLCFastFeatureCount))
+        return TRUE;
+
+    if (EQUAL(pszCap, OLCZGeometries))
+        return TRUE;
+
+    if (EQUAL(pszCap, OLCStringsAsUTF8))
+        return TRUE;
+
+    return FALSE;
+}
+
+/****************************************************************************/
+/*                            CreateField()                                 */
+/****************************************************************************/
+
+OGRErr OGRMiraMonLayer::CreateField(const OGRFieldDefn *poField, int bApproxOK)
+
+{
```

Comment:
same question as for the OLCCreateField capability. can a field be added to a layer with already existing features in it ? if not, then error out

Reply:
Once a feature has been added MiraMon the driver will not accept more Creation of Fields (It could do it but I don't to do it now).
Is the current code enough?

if (!m_bUpdate)
    {
        CPLError(CE_Failure, CPLE_NoWriteAccess,
                 "Cannot create fields on a read-only dataset.");
        return OGRERR_FAILURE;
    }

Reply:
> Is the current code enough?
> 
> if (!m_bUpdate) { CPLError(CE_Failure, CPLE_NoWriteAccess, "Cannot create fields on a read-only dataset."); return OGRERR_FAILURE; }

this covers only the part about erroring out on a file open in read-only mode

To make CreateField() return in error if there are already existing features, you'll likely need to test phMiraMonLayer->TopHeader.nElemCount against 0

Reply:
Done!